athenet/sparsifying/utils/interval.py

```diff
@@ -0,0 +1,284 @@
+"""Intervals in Theano including special functions for sparsifying."""
+
+from athenet.sparsifying.utils.numlike import Numlike
+from theano import function
+from theano import tensor as T
+from theano.ifelse import ifelse
+
+
+class Interval(Numlike):
+    """Theano interval class
+
+    Note: Should be treated as interval type with bounds as Theano nodes.
+    Opetations on Interval create nodes in Theano graph."""
+
+    def __init__(self, lower, upper):
+        """Creates interval.
+
+        :lower: lower bound of Interval to be set
+        :upper: upper bound if Interval to be set
+
+        Note: lower must be lower than upper. It is not being checked."""
+        self.lower = lower
+        self.upper = upper
+
+    def __getitem__(self, at):
+        """Returns specified slice of interval as a interval.
+
+        :at: Coordinates / slice to be taken.
+
+        Does not copy data."""
+        return Interval(self.lower[at], self.upper[at])
+
+    def __setitem__(self, at, other):
+        """Just like Theano set_subtensor function, but as a operator.
+
+        :at: Coordinates / slice to be set.
+        :other: Data to be put at 'at'
+
+        """
+        self.lower = T.set_subtensor(self.lower[at], other.lower)
+        self.upper = T.set_subtensor(self.upper[at], other.upper)
+
+    def __repr__(self):
+        """Standard repr method."""
+        return '[' + repr(self.lower) + ', ' + repr(self.upper) + ']'
+
+    def __str__(self):
+        """"Standard str method."""
+        return '[' + str(self.lower) + ', ' + str(self.upper) + ']'
+
+    def shape(self):
+        """Returns shape of interval. Checks only 'lower' matrix.
+
+        Note: does not require self.upper for computations. Therefore it is not
+        safe, but faster."""
+        return self.lower.shape
+
+    def __add__(self, other):
+        """Returns sum of two intervals.
+
+        :other: Interval to be added."""
+        if isinstance(other, Interval):
+            res_lower = self.lower + other.lower
+            res_upper = self.upper + other.upper
+        else:
+            res_lower = self.lower + other
+            res_upper = self.upper + other
+        return Interval(res_lower, res_upper)
+
+    __radd__ = __add__
+
+    def __sub__(self, other):
+        """Returns difference between two intervals.
+
+        :other: Interval to be subtracted."""
```

Comment:
Code says that other may also be a number.
People should know what this function can do. World needs you. Leave a message.